### PR TITLE
sys/shell/commands/gnrc_icmpv6_echo: test for ICMPv6 reply corruption

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The Linux ping utility has the nice feature that fills the ICMPv6 request payload with a pattern `payload_index & 0xFF`.
Then the ICMPv6 response payload is checked to verify that the pattern is still intact.

This way corrupted messages can be detected.
In the past that revealed some 6lo-fragmentation bugs in Linux when corrupted replies arrived.

This feature is also useful for RIOT, so implement it in RIOTs `ping` command.

### Testing procedure

`ping` should still work with varying lengths.
If packet corruption happened on the data path, this should now be reported.

### CODE REVIEW

**Comment by miri64** [comment]
Mh, is that feature optional on Linux? Because, as documented,

https://github.com/RIOT-OS/RIOT/blob/e63927be955158e0f0bdeee2d3b71ff623ce5694/sys/shell/commands/sc_gnrc_icmpv6_echo.c#L15-L19

the implementation is partly based on the Linux implementation.

**Comment by miri64** [comment]
(So my actual question is: should we make this optional using a flag?)

**Comment by miri64** [comment]
At any rate this most likely will need an adaptation of the corresponding `ShellInteraction`

https://github.com/RIOT-OS/RIOT/blob/5f7a7bc6f152c771e8d8a60edd104931792467b4/dist/pythonlibs/riotctrl_shell/gnrc.py#L485-L499

And its corresponding parser

https://github.com/RIOT-OS/RIOT/blob/5f7a7bc6f152c771e8d8a60edd104931792467b4/dist/pythonlibs/riotctrl_shell/gnrc.py#L122-L137

To deal with the new behavior.

**Comment by benpicco** [comment]
What would we gain by making this optional? The Linux ping utility will always do the corruption check.
For `ShellInteraction`, would I just add `reply["corrupt"] = 1` in case the error line is printed?

**Comment by miri64** [comment]
> What would we gain by making this optional? The Linux ping utility will always do the corruption check.

You wouldn't need to adapt the `riotctrl` stuff and not break the tests ;-).

> For `ShellInteraction`, would I just add `reply["corrupt"] = 1` in case the error line is printed?

The `ShellInteraction` itself probably doesn't need any adoption, since as far as I can see you don't change the command parameters. The parser, however, needs adoption. Since you provide offset information and stuff like that, I would prefer, if that information would be provided in the reply dictionary as well. E.g.

```py
reply["corrupt"] = {"bytes": bytes, "offset": offset}
```

**Comment by benpicco** [comment]
I did a thing.
I don't know how to test this though, so it's untested.

(the reply can either be truncated or corrupted. A truncated reply is not checked for corruption as it's already damaged)

File: dist/pythonlibs/riotctrl_shell/gnrc.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ def __init__(self):
                                   r"( rssi=(?P<rssi>-?\d+) dBm)?"
                                   r"( time=(?P<rtt>\d+.\d+) ms)?"
                                   r"(?P<dup> \(DUP\))?")
+                                  r"(response truncated by (?P<truncated>\d+) byte"
```
Comment (miri64) [inline dist/pythonlibs/riotctrl_shell/gnrc.py:28]: You forgot to move the parenthesis at the end of line above to the last line of your addition. That's what `flake8` is complaining about ;-).

Comment (miri64) [reply dist/pythonlibs/riotctrl_shell/gnrc.py:28]: Would be great to have a test case for this in `vim dist/pythonlibs/riotctrl_shell/tests/test_netif.py`. This way you would have caught the error maybe earlier :-).

**Review by miri64 — CHANGES_REQUESTED** [review]

File: dist/pythonlibs/riotctrl_shell/gnrc.py (hunk 1 above)
Comment (benpicco) [reply dist/pythonlibs/riotctrl_shell/gnrc.py:28]: should be fixed now

Comment (miri64) [reply dist/pythonlibs/riotctrl_shell/gnrc.py:28]: The test case is still missing and `tools-test` is still complaining ...

File: dist/pythonlibs/riotctrl_shell/gnrc.py [hunk 2]
```diff
@@ -24,7 +24,9 @@ def __init__(self):
                                   r"icmp_seq=(?P<seq>\d+) ttl=(?P<ttl>\d+)"
                                   r"( rssi=(?P<rssi>-?\d+) dBm)?"
                                   r"( time=(?P<rtt>\d+.\d+) ms)?"
-                                  r"(?P<dup> \(DUP\))?")
+                                  r"(?P<dup> \(DUP\))?"
+                                  r"(response truncated by (?P<truncated>\d+) byte"
+                                  r"(response corrupt at offset (?P<corrupted>\d+)")
```
Comment (miri64) [inline dist/pythonlibs/riotctrl_shell/gnrc.py:29]: Also I believe there are spaces missing at the end or beginning of the respective strings here.

File: dist/pythonlibs/riotctrl_shell/gnrc.py [hunk 3]
```diff
@@ -24,7 +24,9 @@ def __init__(self):
                                   r"icmp_seq=(?P<seq>\d+) ttl=(?P<ttl>\d+)"
                                   r"( rssi=(?P<rssi>-?\d+) dBm)?"
                                   r"( time=(?P<rtt>\d+.\d+) ms)?"
-                                  r"(?P<dup> \(DUP\))?")
+                                  r"(?P<dup> \(DUP\))?"
+                                  r"( response truncated by (?P<truncated>\d+) byte)"
+                                  r"( response corrupt at offset (?P<corrupted>\d+))")
```
Comment (miri64) [inline dist/pythonlibs/riotctrl_shell/gnrc.py:29]: Now the question remains: why is there a group around this string?

**Review by miri64 — CHANGES_REQUESTED** [review]

**Comment by benpicco** [comment]
How can I test `riotctrl_shell/gnrc.py`?

**Comment by miri64** [comment]
Type `tox` when you are in the `dist/pythonlibs/riotctrl_shell/` directory (you might need to install tox first).

**Comment by miri64** [comment]
(see also the `tools-test / python-tests` Github Actions Workflow).

**Comment by benpicco** [comment]
Thank you! Now I figured out what's going on here.
I re-wrote the error output to be in-line so there is only a single line of output / reply for easier parsing.

**Comment by miri64** [comment]
> Thank you for the hint with `tox`! Now I figured out what's going on here.
> I re-wrote the error output to be in-line so there is only a single line of output / reply for easier parsing.

Now only the `flake8` linter is complaining :-).

**Comment by benpicco** [comment]
Heh, I thought we bumped the limit to 100 characters, but only for C code it seems.

**Comment by miri64** [comment]
> Heh, I thought we bumped the limit to 100 characters, but only for C code it seems.

For C code the coding conventions also [still say 80 chars as soft target](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#general).

**Comment by miri64** [comment]
(but we actually aim for [PEP8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) compliance in the pythonlibs ([not in the test scripts though](https://github.com/RIOT-OS/RIOT/blob/6e0c8db99f9065c06d9455282bca9efca4471ca9/dist/tools/flake8/flake8.cfg#L3)),

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 4]
```diff
@@ -340,8 +376,9 @@ static void _print_reply(_ping_data_t *data, gnrc_pktsnip_t *icmpv6,
         }
         recv_seq = byteorder_ntohs(icmpv6_hdr->seq);
         ipv6_addr_to_str(&from_str[0], from, sizeof(from_str));
+
+        /* we can only calculate RTT if payload was large enough for TX timestamp */
         if (data->datalen >= sizeof(uint32_t)) {
-            triptime = xtimer_now_usec() - unaligned_get_u32(icmpv6_hdr + 1);
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c:344]: How is trip-time set now for the checks below? From what I can tell it is only set in `_check_payload()` which is called _after_ these lines. At least for the first message the RTT will now be wrong.

Comment (benpicco) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c:344]: Ops, forgot to move this when I shuffled the error printing around

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 5]
```diff
@@ -238,6 +236,44 @@ static int _configure(int argc, char **argv, _ping_data_t *data)
...
+}
+
+static bool _check_payload(const void *buf, size_t len,
+                           uint32_t *triptime, uint16_t *corrupt)
+{
+    uint8_t i = 0;
+    const uint8_t *data = buf;
+
+    if (len >= sizeof(uint32_t)) {
+        *triptime = xtimer_now_usec() - unaligned_get_u32(buf);
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c:262]: `now` is now much later fetched. Can this be a problem? Ideally this should come directly after the receive and maybe provided as a parameter to `_check_payload()`.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 6]
```diff
@@ -369,13 +388,42 @@ static void _print_reply(_ping_data_t *data, gnrc_pktsnip_t *icmpv6,
...
+        }
         if (rssi != GNRC_NETIF_HDR_NO_RSSI) {
             printf(" rssi=%"PRId16" dBm", rssi);
         }
         if (data->datalen >= sizeof(uint32_t)) {
             printf(" time=%lu.%03lu ms", (long unsigned)triptime / 1000,
                    (long unsigned)triptime % 1000);
         }
+
+        /* we can only calculate RTT if payload was large enough for TX timestamp */
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c:408]:
```suggestion
        /* we can only calculate RTT if payload was large enough for TX timestamp
         * (the basis of triptime) */
```

**Comment by miri64** [comment]
Using `ping6 -c 100 -i 100 ...` I get similar results for both 

###### master
```
--- fe80::dc1a:a8ff:fe09:45b3 PING statistics ---
100 packets transmitted, 100 packets received, 0% packet loss
round-trip min/avg/max = 0.158/0.438/0.781 ms
```

###### This PR
```
--- fe80::dc1a:a8ff:fe09:45b3 PING statistics ---
100 packets transmitted, 100 packets received, 0% packet loss
round-trip min/avg/max = 0.135/0.449/1.470 ms
```

Tox for `dist/pythonlibs/riotctrl_shell` still passes. Did not bother to test the output for `corrupted` and `truncated` replies. Maybe something with `scapy` can be done there.

**Review by miri64 — APPROVED** [review]
ACK

**Comment by miri64** [comment]
Just to be absolutely sure there is not a problem introduced with `riotctrl_shell`, I also started a [`release-test` Github Action] that uses that python library. If it succeeds (or fails with unrelated errors) we are good to go, I think.

**Comment by miri64** [comment]
> Just to be absolutely sure there is not a problem introduced with `riotctrl_shell`, I also started a ~[`release-test` Github Action](https://github.com/RIOT-OS/RIOT/actions/runs/1119622633)~ [`release-tests` Github Action](https://github.com/RIOT-OS/RIOT/runs/3299701063) (_edit:_ link changed to build correct commit) that uses that python library. If it succeeds (or fails with unrelated errors) we are good to go, I think.

Only the usual suspects (LoRaWAN and a test where the nodes are picked way to far apart) fail. So gooo!

**Pull request merged**